NAS2D/Renderer/Window.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <string>
+#include <utility>
+
+enum class WindowDisplayMode {
+	Windowed
+	,Borderless
+	,Borderless_Fullscreen
+};
+
+struct WindowDesc {
+	//The client dimensions
+	//Window creation process should automatically add space for borders, menus and title bars
+	std::pair<int, int> dimensions{1600,600};
+	std::pair<int, int> position{0,0};
+	WindowDisplayMode mode{WindowDisplayMode::Windowed};
+};
+
+class Window {
+public:
+	virtual ~Window() noexcept {}
+
+	virtual void Open() noexcept = 0;
+	virtual void Close() noexcept = 0;
+
+	virtual void Show() noexcept = 0;
+	virtual void Hide() noexcept = 0;
+	virtual void Unhide() noexcept = 0;
+
+	[[nodiscard]] virtual bool IsOpen() const noexcept = 0;
+	[[nodiscard]] virtual bool IsClosed() const noexcept = 0;
+	[[nodiscard]] virtual bool IsFullscreen() const noexcept = 0;
+	[[nodiscard]] virtual bool IsWindowed() const noexcept = 0;
+
+	/**
+	 * @brief Get the underlying OS-specific window handle.
+	 * @return void* Handle to the window. static_cast to appropriate OS-specific type. For Windows this would be HWND.
+	*/
+	[[nodiscard]] virtual void* NativeHandle() const noexcept = 0;
+
+	/**
+	 * @brief Set the underlying OS-specific window handle.
+	 * @param handle: The handle to the window. For Windows this would be a variable of type HWND.
+	 * @return None
+	*/
+	virtual void NativeHandle(void* handle) noexcept = 0;
```

Comment:
I feel like this should perhaps be split off into a subclass, such as `SdlWindow`. That detail gets a bit specific about how windows are implemented. I understand the importance of having accessor methods like that. Not really sure where the corresponding set method would be used though. I feel like how native window handles are set is normally a private internal detail.

Reply:
That's the intention, but in actual use-case, the user should call this from a `Window` interface pointer through the `Renderer` interface and cast appropriately to avoid OS-specific types bleeding through the interface. For example, DirectX needs a reference to the native window handle in some cases and you'd need to cast to an HWND pointer.

For reference of use see reply to your comment on SetIcon

Reply:
I would much prefer an interface that avoids requiring casts to function correctly. I also want to avoid leaking OS or library specific type information. I know that will be quite a challenge here. I don't really have an answer at this point. This is perhaps one of the reasons why I've been avoiding working on that part of the code. I suspect there is a way to do it though.